docs/proposals/20190704-AutomatedUpgrades.md

```diff
@@ -0,0 +1,197 @@
+
+### Slimming down `rootfs`
+
+To slim down the size of our rootfs, I propose that we remove Kubernetes tarballs entirely and let the `kubelet` download them.
+By doing this we can embrace configurability at the Kubernetes layer.
+The disadvantage to this is that this introduces a number a variables in the number of things that can go wrong.
+Having "approved" tarballs packaged with Talos allows us to run conformance tests against exactly what users will use.
+That being said, I think this is overall a better direction for Talos.
+
+### A New Partition Scheme
+
+With the move to run the `rootfs` in memory, we will remove the need for `ROOT-A` and `ROOT-B` partitions.
+In the case of cloud installs, or where users may want to boot without PXE/iPXE, we will require a boot partition.
+In all cases we require a data partition.
+
+I also propose we rename the `DATA` partition to `EPHEMERAL` to make it clear that this partition should not be used to store application data.
+Our suggestion to users who do want persistent storage on the host is to mount another disk and to use that.
+Talos will make the guarantee to only manage the "owned" block device.
+
+### Performing Upgrades in a Container
+
+The move to running `rootfs` in memory, and thus making the block device ephemeral, and the split of `init` into `init` and `machined`, allows us to take an approach to upgrades that is clean, and relatively low risk.
+The workflow will look roughly like the following:
+
+- an upgrade request is received by `osd` and proxied to `machined`
+- `machined` stops `system` and `k8s.io` services
+- `machined` unmounts all mounts in `/var`
+- `machined` runs the specified installer container
+
+### Automated Upgrades
+
+All of the aformentioned improvements help in the higher goal of automated upgrades.
+Once the above is implemented, upgrades should be simple and self-contained to a machine.
+By running upgrades in containers, we can have very simple operator implementation that can rollout an upgrade across a cluster by simply:
+
+- checking for a new tag
+- orchestrating cluster-wide rollout
+
+The operator will initially coordinate upgrades only, but it will also provide a basis for a number of future automation opportunities.
+
+#### Release Channels
+
+We will maintain a service that allows for the operator to discover when new releases of Talos are made available.
+There will be three "channels" that a machine can subscribe to:
+
+- alpha
+- beta
+- stable
+
+The implementation details around this service and its' interaction with the operator is not in the scope of this proposal.
+
+#### Workflow
+
+In order to upgrade nodes, the operator will subscribe to events for nodes.
+It will keep an in-memory list of all nodes.
+The operator will periodically query the target registry for a list of containers within a time range.
+The time range should start where a previous query has ended.
```

Comment:
The initial query should use the currently installed image's timestamp to decide where to start. This means we should add this as a label.